[core][collective] Support customizing gloo timeout

HollowMan6 opened the pull request:
## Why are these changes needed?

Should be used together with https://github.com/ray-project/pygloo/pull/34 so that the parameters can be properly passed to gloo.

<!-- Please give a short summary of the change and the problem this solves. -->

<!-- For example: "Closes #1234" -->

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jovany-wang reviewed (approved):
LG

jovany-wang commented:
@HollowMan6 BTW, what's your usecase?

HollowMan6 commented:
@jovany-wang Thanks for reviewing! It's for resolving the timeout issue when sending large tensors via gloo with ray collective communication lib.

jovany-wang commented:
@HollowMan6  Are you going to apply this feature into OpenRLHF or something like it?

jovany-wang commented:
Oops, maybe ask for an unit test.

HollowMan6 commented:
> Are you going to apply this feature into OpenRLHF or something like it?

Yes and probably sometime after this release. But actually I think this is widely applicable whenever users think that 30 seconds of default gloo timeout is not enough for them.

> Oops, maybe ask for an unit test.

I added it to `test_gloo_group_isolation.py` and fixed some errors in the test file. By the way, I also make `get_group_handle` public, as it could be useful for users to access the group handle and more fine-grained control.

@jovany-wang
